README.md
<p align="center">
  <img style="width:100%;height:auto; max-width:635px" src="http://pinc.nz/git/twitter/twittermeta.png">
</p>

[![Total Downloads](https://poser.pugx.org/toastnz/twitter-card-meta/downloads)](https://packagist.org/packages/toastnz/twitter-card-meta)
[![License](https://poser.pugx.org/toastnz/twitter-card-meta/license)](https://packagist.org/packages/toastnz/twitter-card-meta)
[![Build Status](https://travis-ci.org/toastnz/twitter-card-meta.svg?branch=master)](https://travis-ci.org/toastnz/twitter-card-meta)
<img src="https://img.shields.io/badge/made_with-%20%20%20%E2%98%95-green.svg">
<img src="https://img.shields.io/badge/made_by-%F0%9F%92%A9%F0%9F%92%A9-dddddd.svg">
<img src="https://img.shields.io/badge/made_for-%F0%9F%92%B5-e6acca.svg">

# Silverstripe Twitter Card Module

Simple Module to add Twitter Cards to your site

## Installation

Get the goodness just like so:

```
 composer require toastnz/twitter-card-meta
```

And ensure you have the following in the `<head>` of your template:

```
 $MetaTags
```

If you are not using `$MetaTags`, you can use the following template code:

```
  <% include TwitterCardMeta %>
```

Step 4: Profit




<p align="center">
  <img style="width:100%;height:auto; max-width:938px" src="http://pinc.nz/git/twitter/settings.png">
</p>


changelog.md
# Change Log
---

## [1.0.0](https://github.com/toastnz/twitter-card-meta/tree/1.0.0) (2017-06-28)

* Disable saving Twitter details to database
* Move to single CMS tab

## [0.2.4](https://github.com/toastnz/twitter-card-meta/tree/0.2.4) (2016-11-28)

* Change from template include to MetaTags extension

---

## [0.2.2](https://github.com/toastnz/twitter-card-meta/tree/0.2.2) (2016-11-11)

* Fix issue where title "New Page" was automatically saved to Twitter Card
* Relax CMS / Framework requirement

---

## [0.2.1](https://github.com/toastnz/twitter-card-meta/tree/0.2.1) (2016-07-01)

* Unit testing

---

## [0.2.0](https://github.com/toastnz/twitter-card-meta/tree/0.2.0) (2016-06-02)

* Added migration task for older versions; moves data from the Page table to the SiteTree table
* General adherence to SilverStripe module standards

---

## [0.1.0](https://github.com/toastnz/twitter-card-meta/tree/0.1.0) (2016-04-22)

* **BREAKING:** Move TwitterCardMeta from Page to SiteTree
* **BUG:** Fix issue where the initial data saved would base off auto-generated Title etc (e.g.; New Home Page)

---

## [0.0.1](https://github.com/toastnz/twitter-card-meta/tree/0.0.1) (2016-04-08)

* Initial development
* Basic module standards


tests/TwitterCardMetaTest.php
<?php

use SilverStripe\Assets\File;
use SilverStripe\SiteConfig\SiteConfig;
use SilverStripe\CMS\Model\SiteTree;
use SilverStripe\Control\Director;
use SilverStripe\Control\Controller;
use SilverStripe\Dev\SapphireTest;

/**
 * Class TwitterCardMetaTest
 *
 * @mixin PHPUnit_Framework_Assert
 */
class TwitterCardMetaTest extends SapphireTest
{
    protected static $fixture_file = 'fixtures/TwitterCardMetaTest.yml';

    public function setUp()
    {
        /** =========================================
         * @var SiteConfig $siteConfig
         * ========================================*/

        parent::setUp();

        $defaultTwitterImage = $this->objFromFixture(File::class, 'default');

        $siteConfig = SiteConfig::current_site_config();
        $siteConfig->setField('DefaultTwitterHandle', 'silverstripe');
        $siteConfig->setField('DefaultTwitterImageID', $defaultTwitterImage->ID);
        $siteConfig->write();
    }

    /**
     * @covers TwitterCardMeta::getCreatorHandle()
     */
    public function testGetCreatorHandle()
    {
        /** =========================================
         * @var TwitterCardMeta|SiteTree $aboutPage
         * ========================================*/

        $aboutPage = $this->objFromFixture(SiteTree::class, 'about');
        $handle = $aboutPage->getCreatorHandle();
        $this->assertEquals('toastnz', $handle);

        $aboutPage = $this->objFromFixture(SiteTree::class, 'home');
        $handle = $aboutPage->getCreatorHandle();
        $this->assertEquals('silverstripe', $handle);
    }

    /**
     * @covers TwitterCardMeta::getFirstImage()
     */
    public function testGetFirstImage()
    {
        /** =========================================
         * @var TwitterCardMeta|SiteTree $aboutPage
         * @var TwitterCardMeta|SiteTree $homePage
         * ========================================*/

        $aboutPage = $this->objFromFixture(SiteTree::class, 'about');
        $firstImage = $aboutPage->getFirstImage();
        $this->assertEquals('assets/Uploads/6868265.gif', $firstImage);

        $homePage = $this->objFromFixture(SiteTree::class, 'home');
        $firstImage = $homePage->getFirstImage();
        $this->assertEquals('', $firstImage, 'Found a result when there should be none');
    }

    /**
     * @covers TwitterCardMeta::onBeforeWrite()
     */
    public function testOnBeforeWrite()
    {
        /** =========================================
         * @var TwitterCardMeta|SiteTree $homePage
         * @var TwitterCardMeta|SiteTree $aboutPage
         * ========================================*/

        /** -----------------------------------------
         * Homepage
         * ----------------------------------------*/

        $homePage = $this->objFromFixture(SiteTree::class, 'home');
        $homePage->write();

        $defaultTwitterImage = $this->objFromFixture(File::class, 'default');

        $this->assertEquals(
            'Cras luctus. Convallis etiam proin urna, consequat nibh vulputate luctus laoreet venenatis vestibulum malesuada vehicula.',
            $homePage->TwitterDescription
        );

        $this->assertEquals('Home', $homePage->TwitterTitle);
        $this->assertEquals('silverstripe', $homePage->TwitterCreator);
        $this->assertEquals('silverstripe', $homePage->TwitterSite);
        $this->assertEquals($defaultTwitterImage->ID, $homePage->TwitterImageID);

        /** -----------------------------------------
         * About page
         * ----------------------------------------*/

        $aboutPage = $this->objFromFixture(SiteTree::class, 'about');
        $aboutPage->write();

        $defaultTwitterImage = $this->objFromFixture(File::class, 'alt');

        $this->assertEquals(
            'Welcome to SilverStripe!',
            $aboutPage->TwitterDescription
        );

        $this->assertEquals('About', $aboutPage->TwitterTitle);
        $this->assertEquals('toastnz', $aboutPage->TwitterCreator);
        $this->assertEquals('silverstripe', $aboutPage->TwitterSite);
        $this->assertEquals($defaultTwitterImage->ID, $aboutPage->TwitterImageID);
    }

    /**
     * @covers TwitterCardMeta::getTwitterImageURL()
     */
    public function testGetTwitterImageURL()
    {
        /** =========================================
         * @var SiteConfig $siteConfig
         * @var Image $defaultTwitterImage
         * @var TwitterCardMeta|SiteTree $homePage
         * @var TwitterCardMeta|SiteTree $aboutPage
         * ========================================*/

        $siteConfig = SiteConfig::current_site_config();

        $defaultTwitterImage = $this->objFromFixture(File::class, 'default');

        $siteConfig->setField('DefaultTwitterHandle', '');
        $siteConfig->setField('DefaultTwitterImageID', 0);
        $siteConfig->write();

        $expectedImageURL = $defaultTwitterImage->CroppedImage(560, 750)->AbsoluteURL;

        /** -----------------------------------------
         * Test without default
         * ----------------------------------------*/

        $homePage = $this->objFromFixture(SiteTree::class, 'home');
        $imageURL = $homePage->getTwitterImageURL();
        $this->assertEquals('', $imageURL);

        $aboutPage = $this->objFromFixture(SiteTree::class, 'about');
        $imageURL = $aboutPage->getTwitterImageURL();
        $this->assertEquals(Controller::join_links(Director::absoluteBaseURL(), 'assets/Uploads/6868265.gif'), $imageURL);

        $aboutPage->setField('TwitterImageID', $defaultTwitterImage->ID);
        $aboutPage->write();
        $imageURL = $aboutPage->getTwitterImageURL();
        $this->assertEquals($expectedImageURL, $imageURL);

        /** -----------------------------------------
         * Test with default
         * ----------------------------------------*/

        $siteConfig->setField('DefaultTwitterImageID', $defaultTwitterImage->ID);
        $siteConfig->write();

        $imageURL = $homePage->getTwitterImageURL();
        $this->assertEquals($siteConfig->DefaultTwitterImage()->CroppedImage(560, 750)->AbsoluteURL, $imageURL);
    }
}


tests/MigrateSiteTreeMetaTaskTest.php
<?php

use SilverStripe\ORM\DataExtension;
use SilverStripe\Dev\SapphireTest;

/**
 * Class MigrateSiteTreeMetaTaskTest
 *
 * @mixin PHPUnit_Framework_Assert
 */
class MigrateSiteTreeMetaTaskTest extends SapphireTest
{
    protected static $fixture_file = 'fixtures/MigrateSiteTreeMetaTaskTest.yml';

    protected $extraDataObjects = array('Page');

    public function setUp()
    {
        parent::setUp();

        // Apply the extension to the Page
        DataExtension::add_to_class('Page', 'TwitterCardMeta');
    }

    public function testRunWithOverwrite()
    {
        $oldHomepage = $this->objFromFixture('Page', 'oldHome');

        $this->assertEquals('Home', $oldHomepage->TwitterTitle);
    }
}


code/tasks/MigrateSiteTreeMetaTask.php
<?php

use SilverStripe\Control\Director;
use SilverStripe\ORM\DB;
use SilverStripe\Core\Convert;
use SilverStripe\Dev\BuildTask;

/**
 * Class MigrateSiteTreeMetaTask
 */
class MigrateSiteTreeMetaTask extends BuildTask {

    protected $title = 'Migrate Twitter metadata from Page to SiteTree';

    protected $description = 'For versions < 0.1.0: Migrates existing data from the Page table to the SiteTree table. Pass the argument/GET var "overwrite=1" to replace existing data.';

    protected $eol;

    protected $hr;

    protected $fields_to_update = array(
        'TwitterSite',
        'TwitterCreator',
        'TwitterTitle',
        'TwitterCardType',
        'TwitterDescription',
        'TwitterImageID'
    );

    public function __construct()
    {
        if (!Director::is_cli()) {
            $this->eol = '<br>';
            $this->hr = '<hr>';
        } else {
            $this->eol = PHP_EOL;
            $this->hr = '--' . PHP_EOL;
        }

        parent::__construct();
    }

    /**
     * @param SS_HTTPRequest $request
     */
    public function run($request)
    {
        /** =========================================
         * @var Page $page
        ===========================================*/

        if (class_exists('Page')) {
            if (Page::has_extension('TwitterCardMeta')) {
                // Should we overwrite?
                $overwrite = $request->getVar('overwrite') ? true : false;

                echo sprintf('Overwrite is %s', $overwrite ? 'enabled' : 'disabled') . $this->eol . $this->eol;

                $pages = Page::get();

                foreach ($pages as $page) {

                    $id = $page->ID;

                    echo $this->hr;
                    echo 'Updating page: ' . $page->Title . $this->eol;

                    foreach ($this->fields_to_update as $fieldName) {
                        $oldData = DB::query("SELECT {$fieldName} FROM Page WHERE ID = {$id}")->column($fieldName);

                        $newData = DB::query("SELECT {$fieldName} FROM SiteTree WHERE ID = {$id}")->column($fieldName);

                        if (!empty($oldData)) {
                            // If new data has been saved and we don't want to overwrite, exit the loop
                            if (!empty($newData) && $overwrite === false) {
                                continue;
                            }
                            $data = Convert::raw2sql($oldData[0]);
                            DB::query("UPDATE SiteTree SET {$fieldName} = '{$data}' WHERE ID = {$id}");
                            DB::query("UPDATE SiteTree_Live SET {$fieldName} = '{$data}' WHERE ID = {$id}");
                        } else {
                            echo 'Field "' . $fieldName . '" empty.' . $this->eol;
                        }
                    }
                }
            }
        }
    }
}


code/extensions/TwitterSiteConfigExtension.php
<?php

use SilverStripe\Assets\Image;
use SilverStripe\Forms\FieldList;
use SilverStripe\Forms\TabSet;
use SilverStripe\AssetAdmin\Forms\UploadField;
use SilverStripe\Forms\HeaderField;
use SilverStripe\Forms\TextField;
use SilverStripe\ORM\DataExtension;

/**
 * Class TwitterSiteConfigExtension
 *
 * @property string DefaultTwitterHandle
 * @property int    DefaultTwitterImageID
 *
 * @method Image DefaultTwitterImage
 */
class TwitterSiteConfigExtension extends DataExtension
{
    private static $db = [
        'DefaultTwitterHandle' => 'Varchar(255)'
    ];

    private static $has_one = [
        'DefaultTwitterImage' => Image::class
    ];

    /**
     * @param FieldList $fields
     */
    public function updateCMSFields(FieldList $fields)
    {
        /** =========================================
         * @var UploadField $twitterImage
        ===========================================*/

        if (!$fields->fieldByName('Root.Metadata')) {
            $fields->addFieldToTab('Root', TabSet::create('Metadata'));
        }

        /** -----------------------------------------
         * Details
         * ----------------------------------------*/

        $twitterImage = UploadField::create('DefaultTwitterImage', 'Default Twitter Card Image');
        $twitterImage->setDescription('Ideal size 560px * 750px');

        $fields->findOrMakeTab('Root.Metadata.Twitter');
        $fields->addFieldsToTab('Root.Metadata.Twitter', [
            HeaderField::create('', 'Twitter Cards'),
            TextField::create('DefaultTwitterHandle', 'Default Twitter Handle'),
            $twitterImage
        ]);

    }

}


code/extensions/TwitterCardMeta.php
<?php

use SilverStripe\Assets\Image;
use SilverStripe\Forms\FieldList;
use SilverStripe\Control\Director;
use SilverStripe\Forms\LiteralField;
use SilverStripe\Forms\TextField;
use SilverStripe\Forms\OptionsetField;
use SilverStripe\Forms\TextareaField;
use SilverStripe\AssetAdmin\Forms\UploadField;
use SilverStripe\Forms\ToggleCompositeField;
use SilverStripe\SiteConfig\SiteConfig;
use SilverStripe\Control\Controller;
use SilverStripe\ORM\DataExtension;

/**
 * Class TwitterCardMeta
 *
 * @property SiteTree $owner
 */
class TwitterCardMeta extends DataExtension
{
    /**
     * @var array
     */
    private static $db = [
        'TwitterSite'        => 'Varchar(160)',
        'TwitterCreator'     => 'Varchar(160)',
        'TwitterTitle'       => 'Varchar(160)',
        'TwitterCardType'    => 'Varchar(160)',
        'TwitterDescription' => 'Varchar(160)'
    ];

    /**
     * @var array
     */
    private static $has_one = [
        'TwitterImage' => Image::class
    ];

    /**
     * @param FieldList $fields
     */
    public function updateCMSFields(FieldList $fields)
    {
        /** =========================================
         * @var TextareaField $address
        ===========================================*/

        if (!$this->owner->TwitterCardType) {
            $this->owner->TwitterCardType = 'summary_large_image';
        }

        $fields->addFieldToTab('Root.Main', ToggleCompositeField::create('Twitter Graph', 'Twitter Card',
            [
                LiteralField::create('', '<h2>&nbsp;&nbsp;&nbsp;Twitter Card <img style="position:relative;top:4px;left 4px;" src="' . Director::absoluteBaseURL() . 'twitter-card-meta/images/twitter.png"></h2>'),
                TextField::create('TwitterCreator', 'Creator Handle')
                    ->setAttribute('placeholder', 'e.g @username')
                    ->setRightTitle('Twitter account name for the author/creator (Will default to site handle)'),
                OptionsetField::create('TwitterCardType', 'Twitter Card Type', [
                    'summary_large_image' => 'summary with large image',
                    'summary'             => 'summary',
                ], 'summary_large_image')
                    ->setRightTitle('Choose which type of twitter card you would like this page to share as.'),
                TextField::create('TwitterTitle', 'Twitter Card Title')
                    ->setAttribute('placeholder', 'e.g Description Of Page Content')
                    ->setRightTitle('Twitter title to to display on the Twitter card'),
                TextareaField::create('TwitterDescription', '')
                    ->setRightTitle('Twitter card description goes here, automatically defaults to the content summary'),
                UploadField::create('TwitterImage', 'Twitter Card Image')
                    ->setRightTitle('Will default too the first image in the WYSIWYG editor or banner image if left blank'),
            ]
        ));
    }

    /**
     * Template helper
     *
     * @return string|null
     */
    public function getCreatorHandle()
    {
        return $this->owner->TwitterCreator ?: SiteConfig::current_site_config()->DefaultTwitterHandle;
    }

    /**
     * Controller logic for returning Twitter card image
     *
     * @return String
     */
    public function getTwitterImageURL()
    {
        if ($this->owner->TwitterImage() && $this->owner->TwitterImage()->exists()) {
            return $this->owner->TwitterImage()->Fit(560, 750)->AbsoluteURL;
        } elseif ($firstImage = $this->getFirstImage()) {
            return Controller::join_links(Director::absoluteBaseURL(), $firstImage);
        } elseif (SiteConfig::current_site_config()->DefaultTwitterImage() && SiteConfig::current_site_config()->DefaultTwitterImage()->exists()) {
            return SiteConfig::current_site_config()->DefaultTwitterImage()->Fit(560, 750)->AbsoluteURL;
        }

        return '';
    }

    /**
     * @return string
     */
    public function getFirstImage()
    {
        $pattern = ' /<img[^>]+ src[\\s = \'"]';
        $pattern .= '+([^"\'>\\s]+)/is';
        if (preg_match($pattern, $this->owner->Content, $match) !== false && !empty($match)) {
            $imageLink = preg_replace('/_resampled\/resizedimage[0-9]*-/', '', $match[1]);
            return (string)$imageLink;
        } else {
            return '';
        }
    }

    /**
     * @param string $tags
     */
    public function MetaTags(&$tags)
    {
        /** =========================================
         * @var TwitterSiteConfigExtension $siteConfig
        ===========================================*/

        $siteConfig = SiteConfig::current_site_config();

        // Type
        if ($this->owner->TwitterCardType) {
            $tags .= sprintf('<meta name="twitter:card" content="%s">', $this->owner->TwitterCardType) . "\n";
        } else {
            $tags .= '<meta name="twitter:card" content="summary_large_image">' . "\n";
        }

        // Site
        $siteHandle = $this->owner->TwitterSite ?: $siteConfig->DefaultTwitterHandle ?: '';
        $tags       .= sprintf('<meta name="twitter:site" content="%s">', $siteHandle) . "\n";

        // Creator
        if ($this->owner->TwitterSite) {
            $tags .= sprintf('<meta name="twitter:creator" content="%s">', $this->owner->TwitterSite) . "\n";
        } else {
            $tags .= sprintf('<meta name="twitter:creator" content="%s">', $siteHandle) . "\n";
        }

        // Title
        if ($this->owner->TwitterTitle) {
            $tags .= sprintf('<meta name="twitter:title" content="%s">', $this->owner->TwitterTitle) . "\n";
        } else {
            $tags .= sprintf('<meta name="twitter:title" content="%s">', $this->owner->Title) . "\n";
        }

        // Description
        if ($this->owner->TwitterDescription) {
            $tags .= sprintf('<meta name="twitter:description" content="%s">', $this->owner->TwitterDescription) . "\n";
        } else {
            $tags .= sprintf('<meta name="twitter:description" content="%s">', $this->owner->dbObject('Content')->FirstParagraph()) . "\n";
        }

        // Image
        $image = $this->owner->getTwitterImageURL();

        $this->owner->extend('updateTwitterImage', $image);

        $tags .= sprintf('<meta name="twitter:image" content="%s">', $image) . "\n";
    }

}
